src/Libraries/Microsoft.Extensions.AI.Abstractions/Utilities/DataContentAsyncEnumerableStream.cs

```diff
@@ -0,0 +1,228 @@
+                _remainingDataOffset += bytesToCopy;
+                bytesRead += bytesToCopy;
+                _position += bytesToCopy;
+            }
+            else
+            {
+                // Special case when the first chunk was skipped and needs to be read
+                if (_position == 0 && _firstDataContent is not null && _firstDataContent.Data.HasValue)
+                {
+                    _remainingData = _firstDataContent.Data.Value.ToArray();
+                    _remainingDataOffset = 0;
+
+                    continue;
+                }
+
+                if (!await _enumerator.MoveNextAsync().ConfigureAwait(false))
+                {
+                    _isCompleted = true;
+                    break;
+                }
+
+                if (!_enumerator.Current.Data.HasValue)
+                {
+                    _isCompleted = true;
+                    break;
+                }
+
+                _remainingData = _enumerator.Current.Data.Value.ToArray();
+                _remainingDataOffset = 0;
+            }
+        }
+
+        return bytesRead;
+    }
+#endif
+
+#if NET8_0_OR_GREATER
+    /// <inheritdoc/>
+    public override async ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken = default)
+    {
+        if (_isCompleted)
+        {
+            return 0;
+        }
+
+        int bytesRead = 0;
+        int totalToRead = buffer.Length;
+
+        while (bytesRead < totalToRead)
+        {
+            // If there's still data in the current chunk
+            if (_remainingDataOffset < _remainingData.Length)
+            {
+                int bytesToCopy = Math.Min(totalToRead - bytesRead, _remainingData.Length - _remainingDataOffset);
+                _remainingData.AsSpan(_remainingDataOffset, bytesToCopy)
+                              .CopyTo(buffer.Span.Slice(bytesRead, bytesToCopy));
+
+                _remainingDataOffset += bytesToCopy;
+                bytesRead += bytesToCopy;
+                _position += bytesToCopy;
+            }
+            else
+            {
+                // If the first chunk was never read, attempt to read it now
+                if (_position == 0 && _firstDataContent is not null && _firstDataContent.Data.HasValue)
+                {
+                    _remainingData = _firstDataContent.Data.Value.ToArray();
+                    _remainingDataOffset = 0;
+                    continue;
+                }
+
+                // Move to the next chunk in the async enumerator
+                if (!await _enumerator.MoveNextAsync().ConfigureAwait(false) ||
+                    !_enumerator.Current.Data.HasValue)
+                {
+                    _isCompleted = true;
+                    break;
+                }
+
+                _remainingData = _enumerator.Current.Data.Value.ToArray();
+                _remainingDataOffset = 0;
+            }
+        }
+
+        return bytesRead;
+    }
```

Comment:
Depending on how the stream is used, it could be nice to override CopyToAsync. That implementation can just enumerate and write out each chunk.

Reply:
I'm not fully sure if we can get much of a benefit as this method also expects a bufferSize, and I would need to be careful how to handle that if each DataContent has a different buffer size (which is being taken care of within the `ReadAsync` override impl).